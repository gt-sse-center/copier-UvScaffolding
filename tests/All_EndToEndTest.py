import copy
import os
import re
import sys

from pathlib import Path
from typing import cast
import pytest
import tomlkit

pytest.register_assert_rewrite("TestHelpers")
import TestHelpers


# ----------------------------------------------------------------------
@pytest.mark.filterwarnings("ignore:Dirty template changes included automatically")
@pytest.mark.parametrize(
    "configuration_info",
    [copy.deepcopy(ci) for ci in TestHelpers.ConfigurationInfo.Generate()],
    ids=[ci.name for ci in TestHelpers.ConfigurationInfo.Generate()],
)
def test_All(configuration_info, copie, snapshot) -> None:
    project_dir = TestHelpers.RunTest(copie, configuration_info.configuration)
    assert project_dir is not None

    results: dict[str, str | None] = {}

    for root_str, directories, filenames in os.walk(project_dir):
        root = Path(root_str)
        if root.parts[-1] in [".git", ".venv"]:
            directories[:] = []
            continue

        relative_path = root.relative_to(project_dir)

        if not filenames:
            results[relative_path.as_posix()] = None
            continue

        for filename in filenames:
            if filename == "uv.lock":
                continue

            # Multiple files are populated with dynamic information that is likely to be
            # different between when the test was run and when this code was written. Replace
            # that dynamic information with placeholders that can be used for testing.
            content = (root / filename).read_text(encoding="utf-8")

            for freeform_string in TestHelpers.freeform_strings:
                content = content.replace(
                    configuration_info.configuration[freeform_string], f"<<{freeform_string}>>"
                )

            python_version = ".".join(str(i) for i in sys.version_info[:2])

            if filename == ".copier-answers.yml":
                content = content.replace(str(Path.cwd()), "<<cwd>>")
                content = re.sub(
                    r"^(?P<whitespace>\s+)_commit: .+?$",
                    lambda match: f"{match['whitespace']}_commit: <<commit>>",
                    content,
                    flags=re.MULTILINE,
                )
            elif filename == ".python-version":
                assert content.strip() == python_version, (content, python_version)
                content = "<<python_version>>"
            elif filename == "pyproject.toml":
                toml_content = tomlkit.loads(content)

                project_section = cast(tomlkit.TOMLDocument, toml_content["project"])

                # project.requires-python
                assert project_section["requires-python"] == f">={python_version}", (
                    project_section["requires-python"],
                    python_version,
                )
                project_section["requires-python"] = ">= <<python_version>>"

                # project.authors

                # Note that it is difficult to populate the authors with placeholder values, as this
                # section will not always be present and sections added to the toml will appear at
                # the end of the file. To ensure consistency between those invocations where author
                # is present (and the sections appears in the middle of the file) and those where
                # author is not initially present (where the section would appear at the end of the
                # file when added), we remove the author section entirely.
                project_section.pop("authors", None)

                # dependency-groups.dev
                dependency_group_section = cast(tomlkit.TOMLDocument, toml_content["dependency-groups"])

                dev_dependencies = cast(list, dependency_group_section["dev"])

                # A regex to extract the dependency name and ignore the version information that
                # comes after it.
                dev_dependency_regex = re.compile(r"^(?P<name>\w+).+?")

                versionless_dev_dependencies: list[str] = []

                for dev_dependency in dev_dependencies:
                    match = dev_dependency_regex.match(dev_dependency)
                    assert match, dev_dependency

                    versionless_dev_dependencies.append(match["name"])

                dependency_group_section["dev"] = sorted(versionless_dev_dependencies)

                content = tomlkit.dumps(toml_content)

            results[(relative_path / filename).as_posix()] = content

    assert results == snapshot
