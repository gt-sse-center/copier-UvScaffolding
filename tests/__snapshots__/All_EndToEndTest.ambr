# serializer version: 1
# name: test_All[]
  dict({
    '.copier-answers.yml': '''
      # Changes here will be overwritten by Copier; DO NOT MODIFY THIS FILE DIRECTLY!
      #
      # This file was generated using copier (https://copier.readthedocs.io/) with the
      # template copier-UvScaffolding (https://github.com/gt-sse-center/copier-UvScaffolding).
      # Additional information (including instructions on how to use this template with copier) is
      # available at https://github.com/gt-sse-center/copier-UvScaffolding.
      #
      
      _commit: <<commit>>
      _src_path: <<cwd>>
      author_email: <<author_email>>
      author_name: <<author_name>>
      github_host: https://github.com
      github_repo_name: <<github_repo_name>>
      github_username: <<github_username>>
      project_name: this_is_the_project_name
      python_package_name: this_is_the_project_name
  
    ''',
    '.github/ISSUE_TEMPLATE/bug_report.md': '''
      ---
      name: Bug report
      about: Create a report to help us improve.
      title: ''
      labels: ''
      assignees: ''
      
      ---
      
      ## Describe the bug
      A clear and concise description of what the bug is.
      
      ## To Reproduce
      Steps to reproduce the behavior:
      1. Go to '...'
      2. Click on '....'
      3. Scroll down to '....'
      4. See error
      
      ## Expected behavior
      A clear and concise description of what you expected to happen.
      
      ## Screenshots
      If applicable, add screenshots to help explain your problem.
      
      ## Environment (please complete the following information):
       - OS: [e.g. iOS]
       - Python Version: [e.g. 3.13, 3.12, 3.11, etc.]
       - dbrownell_DevTools Version [e.g. 1.2.3]
      
      ## Additional context
      Add any other context about the problem here.
  
    ''',
    '.github/ISSUE_TEMPLATE/config.yml': '''
      blank_issues_enabled: false
  
    ''',
    '.github/ISSUE_TEMPLATE/custom.md': '''
      ---
      name: Custom issue template
      about: Create an issue that isn't a bug or feature request.
      title: ''
      labels: ''
      assignees: ''
      
      ---
  
    ''',
    '.github/ISSUE_TEMPLATE/feature_request.md': '''
      ---
      name: Feature request
      about: Suggest an idea for this project.
      title: ''
      labels: ''
      assignees: ''
      
      ---
      
      ## Is your feature request related to a problem? Please describe.
      A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]
      
      ## Describe the solution you'd like
      A clear and concise description of what you want to happen.
      
      ## Describe alternatives you've considered
      A clear and concise description of any alternative solutions or features you've considered.
      
      ## Additional context
      Add any other context or screenshots about the feature request here.
  
    ''',
    '.github/pull_request_template.md': '''
      ## :pencil: Description
      Please include a summary of the change.
      
      ## :gear: Work Item
      Please include link to the corresponding GitHub Issue or Project work item.
      
      ## :movie_camera: Demo
      Please provide any images, GIFs, or videos that show the effect of your changes (if applicable). A picture is worth a thousand words.
  
    ''',
    '.gitignore': '''
      # Python-generated files
      __pycache__/
      *.py[oc]
      build/
      dist/
      wheels/
      *.egg-info
      
      # Virtual environments
      .venv
      .coverage
      .DS_Store
      .vscode
      minisign_key.pri
    ''',
    '.python-version': '''
      3.13
  
    ''',
    'README.md': '',
    'pyproject.toml': '''
      [project]
      name = "this-is-the-project-name"
      description = "Add your description here"
      readme = "README.md"
      authors = [
          { name = "David Brownell", email = "db@davidbrownell.com" }
      ]
      requires-python = ">=3.13"
      dependencies = []
      dynamic = ["version"]
      classifiers = [
          "License :: OSI Approved :: MIT License",
          "Operating System :: MacOS",
          "Operating System :: Microsoft :: Windows",
          "Operating System :: POSIX :: Linux",
          "Programming Language :: Python",
          "Programming Language :: Python :: 3.13",
          "Programming Language :: Python :: 3.12",
          "Programming Language :: Python :: 3.11",
          "Programming Language :: Python :: 3.10",
      ]
      
      [project.urls]
      Homepage = "https://github.com/<<github_username>>/<<github_repo_name>>"
      Documentation = "https://github.com/<<github_username>>/<<github_repo_name>>"
      Repository = "https://github.com/<<github_username>>/<<github_repo_name>>"
      
      [build-system]
      requires = ["hatchling"]
      build-backend = "hatchling.build"
      
      [dependency-groups]
      dev = [
          "pytest-cov>=6.0.0",
          "pytest>=8.3.4",
          "ruff>=0.9.0",
          "autogitsemver>=0.7.8",
      ]
      
      [tool.hatch.version]
      path = "src/this_is_the_project_name/__init__.py"
      
      [tool.pytest.ini_options]
      addopts = "--verbose -vv --capture=no --cov=this_is_the_project_name --cov-fail-under=95.0"
      
      [tool.ruff]
      line-length = 110
      
      [tool.ruff.lint]
      exclude = ["tests/**"]
      
      select = ["ALL"]
      
      ignore = [
          "ANN002", # Missing type annotation for `*args`
          "ANN003", # Missing type annotation for `**kwargs`
          "BLE001", # Do not catch blind exception: `Exception`
          "COM812", # Trailing comma missing
          "D202", # No blank lines allowed after function docstring
          "E501", # Line too long
          "I001", # Import block is un-sorted or un-formatted
          "N802", # Function name `xxx` should be lowercase
          "N999", # Invalid module name
          "S101", # Use of assert detected
          "UP032", # Use f-string instead of `format` call
      ]
      
      [tool.ruff.lint.mccabe]
      max-complexity = 15
      
      [tool.ruff.lint.pylint]
      max-args = 10
      max-branches = 20
      max-returns = 20
  
    ''',
    'src/AutoGitSemVer.yaml': '''
      # This file is used by AutoGitSemVer to configure the semantic version of this package. It indicates
      # that AutoGitSemVer should only look at changes in this directory, its descendants, and files
      # specified in this list when calculating the semantic version for the package.
      additional_dependencies:
        - "../pyproject.toml"
        - "../README.md"
        - "../uv.lock"
  
    ''',
    'src/this_is_the_project_name/__init__.py': '''
      def hello() -> str:
          return "Hello from this-is-the-project-name!"
      
      # Wheel names will be generated according to this value. Do not manually modify this value; instead
      # update it according to committed changes by running this command from the root of the repository:
      #
      #   uv run python -m AutoGitSemVer.scripts.UpdatePythonVersion ./src/this_is_the_project_name/__init__.py ./src
      #
      __version__ = "0.1.0"
  
    ''',
    'src/this_is_the_project_name/py.typed': '',
  })
# ---
